core/templates/components/question-directives/questions-list/questions-list.component.ts

```diff
@@ -273,51 +273,69 @@ export class QuestionsListComponent implements OnInit, OnDestroy {
-        questionId, [
-          {
-            id: this.selectedSkillId,
-            task: 'remove'
-          } as SkillLinkageModificationsArray
-        ]
-      ).then(() => {
-        this.questionsListService.resetPageNumber();
-        this.questionsListService.getQuestionSummariesAsync(
-          this.selectedSkillId, true, true);
-        this.alertsService.addSuccessMessage('Deleted Question');
-        this._removeArrayElement(questionId);
-      });
-    } else {
-      this.allSkillSummaries.forEach((summary) => {
-        if (summary.getDescription() === skillDescription) {
-          this.editableQuestionBackendApiService.editQuestionSkillLinksAsync(
-            questionId, [
-              {
-                id: summary.getId(),
-                task: 'remove'
-              } as SkillLinkageModificationsArray
-            ]
-          ).then(() => {
-            this.questionsListService.resetPageNumber();
-            this.questionsListService.getQuestionSummariesAsync(
-              this.selectedSkillId, true, true);
-            this.alertsService.addSuccessMessage('Deleted Question');
-            this._removeArrayElement(questionId);
-          });
+    let numberOfQuestions = this.questionSummariesForOneSkill.length;
+    await this.skillBackendApiService
+      .getTopicNamesWithGivenSkillAssignedForDiagnosticTest(
+        this.selectedSkillId).then((topicNames) => {
+        if ((topicNames.length > 0) && (
+          numberOfQuestions <=
+            AppConstants.MINIMUM_QUESTION_COUNT_FOR_A_DIAGNOSTIC_TEST_SKILL)) {
+          let alertMessge = (
+            'This skill is used as the only diagnostic test skill in the ' +
+            'following topics: ' + topicNames.join(', ') + '. Please remove ' +
+            'this skill from the diagnostic tests of those topics first.'
+          );
+          this.alertsService.addInfoMessage(alertMessge, 7000);
+          questionDeletionIsAllowed = false;
         }
       });
-    }
+    return questionDeletionIsAllowed;
+  }
+
+  removeQuestionSkillLinkAsync(questionId: string, skillId: string): void {
+    this.editableQuestionBackendApiService.editQuestionSkillLinksAsync(
+      questionId, [
+        {
+          id: skillId,
+          task: 'remove'
+        } as SkillLinkageModificationsArray
+      ]
+    ).then(() => {
+      this.questionsListService.resetPageNumber();
+      this.questionsListService.getQuestionSummariesAsync(
+        this.selectedSkillId, true, true);
+      this.alertsService.addSuccessMessage('Deleted Question');
+      this._removeArrayElement(questionId);
+    });
+  }
+
+  deleteQuestionFromSkill(
+      questionId: string, skillDescription: string): void {
+    this.isQuestionDeletionAllowed().then((response) => {
+      if (!response) {
+        return;
+      }
+
+      this.deletedQuestionIds.push(questionId);
+      if (this.allSkillSummaries.length === 0) {
```

Comment:
I'm a little confused about this. What is "allSkillSummaries" -- what does it represent? If there are no skills then is there actually a skill to remove the question from?

After the question is removed, are we updating the list of skills? Maybe the naming could be clearer.

Reply:
> What is "allSkillSummaries" -- what does it represent?

Removal of question can be done from two places
- Topic editor question tab, which contains all the skill summaries associated with a topic.
- Skill editor question tab, in which allSkillSummaries.length will be zero because only a single skill is considered here.

Reply:
[Skill question tab](https://github.com/oppia/oppia/blob/9980b23608f22a227ea53adff5312b8d4dbfc342/core/templates/pages/skill-editor-page/questions-tab/skill-questions-tab.directive.html#L3)

[Topic question tab](https://github.com/oppia/oppia/blob/9980b23608f22a227ea53adff5312b8d4dbfc342/core/templates/pages/topic-editor-page/questions-tab/topic-questions-tab.component.html#L16)


From the above two references, it is shown that in the skill question tab this field is passed as an empty list.

Regarding renaming or restructuring, I think the changes which will be required in the topic-question-tab files and skill-question-tab files are not associated with the diagnostic test so can I create an issue for now, so that it can be handled independently?

Reply:
I thought about it, and initially I leaned towards yes, but the problem I have is that I don't understand the code as written so I can't reason through its correctness (insofar as the diagnostic test is concerned).

So I think some limited renaming would help. I would try not to start refactoring other pages unless it's absolutely necessary, but at a minimum this variable needs to have a name/description that accurately describes what it represents and that makes sense in the context of this file. Is that reasonable?

Reply:
In the updated implementation, allSkillSummaries are not required as a flag for differentiating topic-editor-question-tab and skill-editor-question-tab. As because removeQuestion method should be expected to work in the same fashion for both of the tabs and in reality, the allSkillSummaries is not potentially responsible for segregating the two tabs. Thus I have removed its implementation and made the code simpler.

Please take a look.